File: Medium/49. Group Anagrams.py
class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        anagram = {}
        key: str = ''
        for word in strs:
            key = f'{sorted(word)}'
            if key in anagram:
                anagram[key].append(word)
            else:
                anagram[key] = [word]
        return list(anagram.values())
    

class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        anagram = {}
        key: str = ''
        freq: list[int] = [0 for _ in range(128)]
        for word in strs:
            for letter in word:
                freq[ord(letter)] += 1
            for i in range(128):
                if freq[i] > 0:
                    key += f'{freq[i]}{chr(i)}'

            if key in anagram:
                anagram[key].append(word)
            else:
                anagram[key] = [word]
            freq = [0 for _ in range(128)]
            key = ''
        return list(anagram.values())

File: Medium/49. Group Anagrams.cs
public class Solution
{
    public IList<IList<string>> GroupAnagrams(string[] strs)
    {
        IList<IList<string>> res = [];
        int[] freq = new int[128];
        Array.Clear(freq, 0, freq.Length);
        Dictionary<string, IList<string>> hash = new Dictionary<string, IList<string>>();
        foreach (string item in strs)
        {
            string key = "";
            Array.Clear(freq, 0, freq.Length);
            foreach (char letter in item)
            {
                freq[(int)letter]++;
            }
            for (int i = 0; i < 128; i++)
            {
                if (freq[i] > 0)
                {
                    key += $"{freq[i]}{(char)i}";
                }
            }
            Console.WriteLine(key);
            if (hash.ContainsKey(key))
            {
                hash[key].Add(item);
            }
            else
            {
                hash[key] = new List<string>() { item };
            }
        }

        foreach (IList<string> item in hash.Values)
        {
            res.Add(item);
        }

        return res;
    }
}